### PR TITLE
reduce memory allocation in component update core codepath

### PR DESCRIPTION
**Pull request by ngokevin**
**Description:**

Rather than create a new object each time to create a new `oldData`, just have two objects representing `oldData` and `data`. Also when updating, rather than create a new `data` object, have a `newData` object that is used for holding the new data which then gets copied into `data`. So in total, only use three objects: `oldData`, `data`, `newData`.

**Changes proposed:**
- Adapt `style-attr` implementation to be able to pass in an object for reuse instead of creating a new one each time. Also remove lots of callbacks, `Object.keys()` array creations from their code.  Also tweaked to do prettier stringifies. We manage this forked dependency now. 
- Use object pooling in core component code to reduce object allocations. Recycle them when component is removed.
  - 1. Remove object allocations on each update for deep cloning data to `oldData`.
  - 2. Remove object allocation on initialization for initial `oldData`.
  - 3. Remove object allocation on each update during `buildData` for default value copying.  
  - 4. Remove object alloaction during `buildData`, was using `Object.keys`.
- Add some helper properties to component (`isSingleProp`, `isSinglePropObject`, `isObjectBased`)
- Update our `styleParser` utils a bit to not allocate object + array + use callbacks to convert keys to camelCase, and to cache regexes.

Updated to not delete the keys when recycling an object, but just set every key to `undefined`. This means object pooling best used when objects share a common schema, in case we iterate over keys. So I create different object pools for each type of component, since their object keys will be the same. I also update schema parser to ignore undefined keys.

## Before

screen shot 2017-10-03 at 7 16 43 pm

## After

**Improvements:** See the `updateProperties` and `updateCachedAttrValue` no longer on the list of frequent memory allocators. 

screen shot 2017-10-03 at 7 17 57 pm

- [x] Still need to polish/comment a bit more (and try to remove the clones for default values). But up and working again.
- [x] Tests

### CODE REVIEW

**Comment by ngokevin** [comment]
Ready to go. Did another test with the /performance/animation-set-attribute stress test to measure memory. These recordings were taken over **4 seconds**. We save the following allocation calls:

- buildData (component) (~400KB)
- getKeyValueChunks (style-parser) (~160KB)
- cloneData (~80KB)
- updateProperties (~28KB)

With the particular stress test, it does lose out on ~5FPS (20FPS vs 15FPS), but it's also an unrealistic example since we wouldn't use setAttributes that hard and often for animations. Dodging GC is more important because GC will lose you entire frames.

# Before 

before

# After

after

**Comment by arpu** [comment]
@ngokevin i would like test this change, how do you generate the memory statistics?

File: src/components/camera.js [hunk 1]
```diff
@@ -79,6 +79,11 @@ module.exports.Component = registerComponent('camera', {
       // Camera disabled. Set camera to another camera.
       system.disableSpectatorCamera();
     }
+
+    this.savedPose = {
+      position: el.getAttribute('position'),
```
Comment (dmarcos) [inline src/components/camera.js]: I think we have to allocate new objects here. You will be holding a reference to the entity shared objects that will change.

Comment (ngokevin) [reply src/components/camera.js]: I just removed it, it wasn't used.

**Comment by ngokevin** [comment]
browser memory tools

**Comment by dmarcos** [comment]
It would be nicer if this PR could
be broken in smaller PRs. It touches a lot
of files and it’s hard to follow and see implications

File: src/components/obj-model.js [hunk 2]
```diff
@@ -21,12 +21,16 @@ module.exports.Component = registerComponent('obj-model', {
   update: function () {
     var data = this.data;
     if (!data.obj) { return; }
-    this.remove();
+    this.resetMesh();
```
Comment (dmarcos) [inline src/components/obj-model.js]: Unrelated to the purpose of the PR?

File: src/components/material.js [hunk 3]
```diff
@@ -53,10 +53,16 @@ module.exports.Component = registerComponent('material', {
   },
 
   updateSchema: function (data) {
-    var newShader = data.shader;
-    var currentShader = this.data && this.data.shader;
-    var shader = newShader || currentShader;
-    var schema = shaders[shader] && shaders[shader].schema;
+    var currentShader;
+    var newShader;
```
Comment (dmarcos) [inline src/components/material.js]: unrelated to PR?

File: src/components/geometry.js [hunk 4]
```diff
@@ -61,12 +61,10 @@ module.exports.Component = registerComponent('geometry', {
 
   /**
    * Update geometry component schema based on geometry type.
-   *
-   * @param {object} data - New data passed by Component.
    */
   updateSchema: function (data) {
+    var currentGeometryType = this.oldData && this.oldData.primitive;
```
Comment (dmarcos) [inline src/components/geometry.js]: unrelated to PR?

**Comment by dmarcos** [comment]
styleParser.js and component.js changes could be separate PRs

File: src/components/obj-model.js (hunk 2 above)
Comment (ngokevin) [reply src/components/obj-model.js]: fixes a test

File: src/components/material.js (hunk 3 above)
Comment (ngokevin) [reply src/components/material.js]: fixes test, checks for data / oldData existence

File: src/components/geometry.js (hunk 4 above)
Comment (ngokevin) [reply src/components/geometry.js]: yes, it checks for oldData rather than data

**Comment by ngokevin** [comment]
It's hard and effort to break up. They both relate to the object pooling and both tie together. There's only two major changes:

- remove allocations from style parser by passing an object in (managed with object pool)
- remove allocations from component.js by removing the cloning (managed with object pool)

I detail listed all the changes in the original description

**Comment by arpu** [comment]
this breaks my little app with 

body.js:122 Uncaught TypeError: Cannot read property 'toUpperCase' of undefined at initBody (body.js:122) at HTMLElement.<anonymous> (a-node.js:300) at a-node.js:131 initBody @ body.js:122 emit @ a-node.js:300 (anonymous) @ a-node.js:131 Promise.then (async) load @ a-node.js:128 load @ a-entity.js:255 (anonymous) @ a-scene.js:591 setTimeout (async) play @ a-scene.js:590 attachedCallbackPostCamera @ a-scene.js:128 (anonymous) @ a-scene.js:102 emit @ a-node.js:300 checkUserCamera @ camera.js:98 (anonymous) @ camera.js:58 emit @ a-node.js:300 setObject3D @ a-entity.js:166 init @ camera.js:28 updateProperties @ component.js:315 updateComponent @ a-entity.js:489 updateComponents @ a-entity.js:463 (anonymous) @ a-entity.js:259 (anonymous) @ a-node.js:130 Promise.then (async) load @ a-node.js:128 load @ a-entity.js:255 (anonymous) @ a-entity.js:84 emit @ a-node.js:300 (anonymous) @ a-node.js:131 Promise.then (async) load @ a-node.js:128 load @ a-assets.js:83 (anonymous) @ a-assets.js:70 setTimeout (async) value @ a-assets.js:66


https://github.com/donmccurdy/aframe-physics-system/blob/master/src/components/body/body.js#L64

**Comment by ngokevin** [comment]
can you log this.data, and more information if you're updating it, initializing it...a live example would be great

**Comment by arpu** [comment]
here is a live example https://codesandbox.io/s/5k1wxylxkx

**Comment by ngokevin** [comment]
Thanks, possible to have one with the physics as an unminified project file so I can poke around in it?

**Comment by arpu** [comment]
https://codesandbox.io/s/vv6jm1l06l

with unminified physics

**Comment by arpu** [comment]
@ngokevin  should i open a new bugreport?

**Comment by ngokevin** [comment]
thanks sure
